### ISSUE TITLE
linchpin validate fails when using 'type' instead of 'role'

### ISSUE THREAD

**Issue by herlo** [CentOS-PaaS-SIG/linchpin#677]
In the old version of the topology schema 'type' was an allowed value. When running linchpin validate against it, an error occurs.

```
$ linchpin -p /tmp/workspace/PinFile validate 
Traceback (most recent call last):
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/bin/linchpin", line 11, in <module>
    load_entry_point('linchpin', 'console_scripts', 'linchpin')()
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/herlo/Projects/centos/git/linch-pin/linchpin/shell/__init__.py", line 525, in validate
    old_schema=old_schema)
  File "/home/herlo/Projects/centos/git/linch-pin/linchpin/cli/__init__.py", line 452, in lp_validate
    return self.do_validation(prov_data, old_schema=old_schema)
  File "/home/herlo/Projects/centos/git/linch-pin/linchpin/__init__.py", line 822, in do_validation
    self._validate_topology(topology_data)
  File "/home/herlo/Projects/centos/git/linch-pin/linchpin/__init__.py", line 433, in _validate_topology
    if not v.validate(document):
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 877, in validate
    self.__validate_definitions(definitions, field)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 940, in __validate_definitions
    result = validate_rule(rule)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 922, in validate_rule
    return validator(definitions.get(rule, None), field, value)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 1234, in _validate_schema
    self.__validate_schema_sequence(field, schema, value)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 1259, in __validate_schema_sequence
    update=self.update, normalize=False)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 877, in validate
    self.__validate_definitions(definitions, field)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 940, in __validate_definitions
    result = validate_rule(rule)
  File "/home/herlo/Projects/virtualenvs/v1.6.1-release/lib/python2.7/site-packages/cerberus/validator.py", line 922, in validate_rule
    return validator(definitions.get(rule, None), field, value)
  File "/home/herlo/Projects/centos/git/linch-pin/linchpin/utils/validator.py", line 23, in _validate_anyof
    if value['role'] not in definition['schema']['role']['allowed']:
KeyError: 'role'
```

The PinFile looks like this:

```
---
my-target:
  topology:
    topology_name: 'slave-node'
    resource_groups:
      -
        resource_group_name: 'slave-node'
        resource_group_type: 'openstack'
        resource_definitions:
          - name: 'jslave'
            flavor: 'm1.xlarge'
            type: 'os_server'
            image: 'rhel-7.4-server-x86_64-released'
            count: 1
            keypair: 'ci-ops-central'
            networks:
              - 'QE-jenkins-integration'
        credentials:
          filename: 'ci-rhos.yml'
          profile: 'ci-rhos'
  layout:
    inventory_layout:
      hosts:
        slave-node:
          count: 1
          host_groups:
            - rhel7
            - certificate_authority
            - repositories
            - jenkins_slave
```

Changing `type` to `role` and the PinFile validates.

**Comment by 14rcole** [comment]
As far as I know, the word "type" is a reserved keyword in cerberus

**Comment by 14rcole** [comment]
Sorry, didn't mean to close the issue. I'm working on a fix now

**Comment by herlo** [comment]
I think a simple fix would be to validate *after* converting the topology from the old to the new format. Unless you have a better solution?

**Comment by 14rcole** [comment]
Right now we attempt to validate under the new format, then convert if that doesn't work.  We can convert everything by default, with one caveat: `linchpin validate` currently tells the user under which schema the topology validates.  We would no longer be able to do that.

We could also just try to catch any KeyErrors in the initial validation attempt and convert the topology just as we would with a SchemaError.  If the topology is not valid under the old schema however, the errors in the schema would have to be given to the user relative to the old schema, not the new schema. Also, if we wanted to add a flag to only validate under the new schema (to help users convert), either option would make that quite difficult

**Comment by herlo** [comment]
Fixed by #680
